draft-ietf-quic-transport.md

```diff
@@ -1943,6 +1943,11 @@ the client during connection establishment with a Retry packet (see
 {{validate-retry}}) or in a previous connection using the NEW_TOKEN frame (see
 {{validate-future}}).
 
+Until the path is validated, a server SHOULD pad the Initial, Handshake, 1-RTT
```

Comment:
You need to pad Handshake packets, if you consistently pad Initials: Without the ServerHello, the client won't be able to decrypt the Handshake packet.

Reply:
This is necessary to prevent the ping-pong that is being described in [the issue](https://github.com/quicwg/base-drafts/issues/4183#issue-715343219) which is caused by the receipt of a Handshake packet triggering a retransmission.

Reply:
If you pad all Initial packets (meaning: all datagrams containing Initial), you don't need to pad Handshake packets. An undecryptable packet doesn't reset any timers, so if won't lead to the ping-pong you describe.

This is easier to implement (for me at least), since I can use the same logic for packing packets on the client and server side: Pad all packets that contain an Initial.

Reply:
I agree with @marten-seemann. If the first flight pads in both directions, ACKs of those packets should be enough to validate the 1200 MTU requirement, and further packets shouldn't be required to pad, right?

Reply:
@marten-seemann @nibanks I'm not sure if I agree.

Let me start by pointing out the basics that I think is true, before I dive into the details of the recovery draft:
* When the path MTU is below 1200 bytes, it is better to make sure that none of the packets reach the client rather than letting some pass through, because client might extend their timeout if they receive some packets from the server.
* Machinery defined in the transport draft should not depend on the specifics of the recovery draft.
* What we are trying to provide is a recommendation (SHOULD), and that should be the best practice rather than the minimum that an endpoint is supposed to do. And because it is merely a recommendation, endpoints can adopt alternative strategies.

Regarding the specifics of the recovery draft, [the second paragraph of the recovery draft](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#section-6.2.3-2) states that an server MAY retransmit CRYPTO data before the PTO timer expires, when it receives an Initial packet containing duplicate CRYPTO data.

The question is if this paragraph prohibits servers from retransmitting Handshake packets (or 1-RTT packets) containing CRYPTO data. If such action is prohibited, then we will not have ping-pong. If such action is _not_ prohibited, we might see ping-ping if we only pad datagrams carrying Initial packets. That is because if only the Handshake packets (or the 1-RTT packets) reach the client, the client is explicitly allowed to retransmit Initial packets containing CRYPTO data.

My understanding is that such action is _not_ prohibited. In fact, I would argue that the intent of this section is to allow such an action.

Consider the case where the server sent three datagrams in its first flight: 1) a full-sized datagram carrying an Initial and a Handshake packet, 2) a full-sized datagram carrying a Handshake packet, 3) a half-sized datagram carrying a Handshake packet.

When the server receives an Initial carrying duplicate CRYPTO data, the draft states that it can assume that the Initial that the server sent has been lost. What does that exactly mean? The natural assumption would be that either or both of datagrams _2_ and _3_ reached the client, but _1_ did not. It is far less likely that the client received all the three datagrams, but there was a bit flip in the Initial packet, and that the client is buffering the Handshake packets it was unable to decrypt.

Then, why should a server limit the retransmission to only Initial data?

To summarize, my understanding is that section 6.2.3 of the recovery draft suggests that the server MAY retransmit Handshake packets when receiving an Initial packet that contain duplicate CRYPTO data. Therefore, there's the possibility that we would see a ping-pong even with a server that is following the recommendations in the recovery draft, unless it pads all packets that it sends until the path is validated.